File: Projects/FinSeAPDF/PDFBox/pdmodel/graphics/color/ColorSpaceLab.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.apache.pdfbox.pdmodel.graphics.color;

import java.awt.color.ColorSpace;

import org.apache.pdfbox.pdmodel.common.PDRange;

/**
 * This class represents a CalRGB color space.
 * 
 * The color conversion uses the algorithm described on wikipedia.
 * 
 * The blackpoint isn't used, as I can't find any hint how to do that.
 * 
 */
public class ColorSpaceLab extends ColorSpace
{
    private static final long serialVersionUID = -5769360600770807798L;

    private PDTristimulus whitepoint = null;
    // TODO unused??
    private PDTristimulus blackpoint = null;
    private PDRange aRange = null;
    private PDRange bRange = null;

    /**
     * Default Constructor.
     * 
     */
    public ColorSpaceLab()
    {
        super(ColorSpace.TYPE_3CLR, 3);
    }

    /** 
     * Constructor.
     * 
     * @param whitept whitepoint values
     * @param blackpt blackpoint values
     * @param a range for value a 
     * @param b range for value b
     */
    public ColorSpaceLab(PDTristimulus whitept, PDTristimulus blackpt, PDRange a, PDRange b)
    {
        this();
        whitepoint = whitept;
        blackpoint = blackpt;
        aRange = a;
        bRange = b;
    }

    /**
     * Clip the given value to the given range.
     * 
     * @param x
     *            the given value to be clipped
     * @param range
     *            the range to be used to clip the value to
     * 
     * @return the clipped value
     */
    private Single clipToRange(Single x, PDRange range)
    {
        return Math.min(Math.max(x, range.getMin()), range.getMax());
    }

    private static final Single VALUE_6_29 = 6 / 29;
    private static final Single VALUE_4_29 = 4 / 29;
    private static final Single VALUE_108_841 = 108 / 841;
    private static final Single VALUE_841_108 = 841 / 108;
    private static final Single VALUE_216_24389 = 216 / 24389;

    private Single calculateStage2ToXYZ(Single value)
    {
        if (value >= VALUE_6_29)
        {
            return (Single) Math.pow(value, 3);
        }
        else
        {
            return VALUE_108_841 * (value - VALUE_4_29);
        }
    }

    private Single calculateStage2FromXYZ(Single value)
    {
        if (value >= VALUE_216_24389)
        {
            return (Single) Math.pow(value, 1 / 3);
        }
        else
        {
            return VALUE_841_108 * value + VALUE_4_29;
        }
    }

    /**
     * {@inheritDoc}
     * 
     */
    public Single() toRGB(Single() colorvalue)
    {
        ColorSpace colorspaceXYZ = ColorSpace.getInstance(CS_CIEXYZ);
        return colorspaceXYZ.toRGB(toCIEXYZ(colorvalue));
    }

    /**
     * {@inheritDoc}
     * 
     */
    public Single() fromRGB(Single() rgbvalue)
    {
        ColorSpace colorspaceXYZ = ColorSpace.getInstance(CS_CIEXYZ);
        return fromCIEXYZ(colorspaceXYZ.fromRGB(rgbvalue));
    }

    /**
     * {@inheritDoc}
     * 
     */
    public Single() toCIEXYZ(Single() colorvalue)
    {
        Single a = colorvalue(1);
        if (aRange IsNot Nothing)
        {
            // clip the a value to the given range
            a = clipToRange(a, aRange);
        }
        Single b = colorvalue(2);
        if (bRange IsNot Nothing)
        {
            // clip the b value to the given range
            b = clipToRange(b, bRange);
        }
        Single m = (colorvalue(0) + 16) / 116;
        Single l = m + (a / 500);
        Single n = m - (b / 200);

        Single x = whitepoint.getX() * calculateStage2ToXYZ(l);
        Single y = whitepoint.getY() * calculateStage2ToXYZ(m);
        Single z = whitepoint.getZ() * calculateStage2ToXYZ(n);
        return new Single() { x, y, z };
    }

    /**
     * {@inheritDoc}
     * 
     */
    public Single() fromCIEXYZ(Single() colorvalue)
    {
        Single x = calculateStage2FromXYZ(colorvalue(0) / whitepoint.getX());
        Single y = calculateStage2FromXYZ(colorvalue(1) / whitepoint.getY());
        Single z = calculateStage2FromXYZ(colorvalue(2) / whitepoint.getZ());

        Single l = 116 * y - 116;
        Single a = 500 * (x - y);
        Single b = 200 * (y - z);
        if (aRange IsNot Nothing)
        {
            // clip the a value to the given range
            a = clipToRange(a, aRange);
        }
        if (bRange IsNot Nothing)
        {
            // clip the b value to the given range
            b = clipToRange(b, bRange);
        }
        return new Single() { l, a, b };
    }

}

File: Projects/DMDPDF/PDFBox/pdmodel/graphics/color/ColorSpaceLab.vb
Imports FinSeA.Drawings
Imports FinSeA.org.apache.pdfbox.pdmodel.common

Namespace org.apache.pdfbox.pdmodel.graphics.color

    '   /**
    '* This class represents a CalRGB color space.
    '* 
    '* The color conversion uses the algorithm described on wikipedia.
    '* 
    '* The blackpoint isn't used, as I can't find any hint how to do that.
    '* 
    '*/
    Public Class ColorSpaceLab
        Inherits ColorSpace
        'private static final long serialVersionUID = -5769360600770807798L;

        Private whitepoint As PDTristimulus = Nothing
        ' TODO unused??
        Private blackpoint As PDTristimulus = Nothing
        Private aRange As PDRange = Nothing
        Private bRange As PDRange = Nothing

        '/**
        ' * Default Constructor.
        ' * 
        ' */
        Public Sub New()
            MyBase.New(CSEnum.TYPE_3CLR, 3)
        End Sub

        '/** 
        ' * Constructor.
        ' * 
        ' * @param whitept whitepoint values
        ' * @param blackpt blackpoint values
        ' * @param a range for value a 
        ' * @param b range for value b
        ' */
        Public Sub New(ByVal whitept As PDTristimulus, ByVal blackpt As PDTristimulus, ByVal a As PDRange, ByVal b As PDRange)
            Me.New()
            whitepoint = whitept
            blackpoint = blackpt
            aRange = a
            bRange = b
        End Sub

        '/**
        ' * Clip the given value to the given range.
        ' * 
        ' * @param x
        ' *            the given value to be clipped
        ' * @param range
        ' *            the range to be used to clip the value to
        ' * 
        ' * @return the clipped value
        ' */
        Private Function clipToRange(ByVal x As Single, ByVal range As PDRange) As Single
            Return Math.Min(Math.Max(x, range.getMin()), range.getMax())
        End Function

        Private Const VALUE_6_29 = 6 / 29
        Private Const VALUE_4_29 = 4 / 29
        Private Const VALUE_108_841 = 108 / 841
        Private Const VALUE_841_108 = 841 / 108
        Private Const VALUE_216_24389 = 216 / 24389

        Private Function calculateStage2ToXYZ(ByVal value As Single) As Single
            If (value >= VALUE_6_29) Then
                Return Math.Pow(value, 3)
            Else
                Return VALUE_108_841 * (value - VALUE_4_29)
            End If
        End Function

        Private Function calculateStage2FromXYZ(ByVal value As Single) As Single
            If (value >= VALUE_216_24389) Then
                Return Math.Pow(value, 1 / 3)
            Else
                Return VALUE_841_108 * value + VALUE_4_29
            End If
        End Function

        Public Overrides Function toRGB(ByVal colorvalue As Single()) As Single()
            Dim colorspaceXYZ As ColorSpace = ColorSpace.getInstance(CSEnum.CS_CIEXYZ)
            Return colorspaceXYZ.toRGB(toCIEXYZ(colorvalue))
        End Function

        Public Overrides Function fromRGB(ByVal rgbvalue As Single()) As Single()
            Dim colorspaceXYZ As ColorSpace = ColorSpace.getInstance(CSEnum.CS_CIEXYZ)
            Return fromCIEXYZ(colorspaceXYZ.fromRGB(rgbvalue))
        End Function

        Public Overrides Function toCIEXYZ(ByVal colorvalue As Single()) As Single()
            Dim a As Single = colorvalue(1)
            If (aRange IsNot Nothing) Then
                ' clip the a value to the given range
                a = clipToRange(a, aRange)
            End If
            Dim b As Single = colorvalue(2)
            If (bRange IsNot Nothing) Then
                ' clip the b value to the given range
                b = clipToRange(b, bRange)
            End If
            Dim m As Single = (colorvalue(0) + 16) / 116
            Dim l As Single = m + (a / 500)
            Dim n As Single = m - (b / 200)

            Dim x As Single = whitepoint.getX() * calculateStage2ToXYZ(l)
            Dim y As Single = whitepoint.getY() * calculateStage2ToXYZ(m)
            Dim z As Single = whitepoint.getZ() * calculateStage2ToXYZ(n)
            Return New Single() {x, y, z}
        End Function

        Public Overrides Function fromCIEXYZ(ByVal colorvalue As Single()) As Single()
            Dim x As Single = calculateStage2FromXYZ(colorvalue(0) / whitepoint.getX())
            Dim y As Single = calculateStage2FromXYZ(colorvalue(1) / whitepoint.getY())
            Dim z As Single = calculateStage2FromXYZ(colorvalue(2) / whitepoint.getZ())

            Dim l As Single = 116 * y - 116
            Dim a As Single = 500 * (x - y)
            Dim b As Single = 200 * (y - z)
            If (aRange IsNot Nothing) Then
                ' clip the a value to the given range
                a = clipToRange(a, aRange)
            End If
            If (bRange IsNot Nothing) Then
                ' clip the b value to the given range
                b = clipToRange(b, bRange)
            End If
            Return New Single() {l, a, b}
        End Function

    End Class

End Namespace
